### PR TITLE
Fixing PhysicalFilesWatcher path maintenance (Issue #71386)

### LINKED ISSUE #71386: ConfigurationBuilder.AddJsonFile can't fire OnChange event using relative paths
### Description

when using AddJsonFile("./api.json", true, true) it will not fire onChange event but using AddJsonFile("api.json", true, true) it will fire

### Reproduction Steps

var builder = new ConfigurationBuilder();
ConfigBuilder = builder.AddJsonFile("./api.json", true, true).Build();
ChangeToken.OnChange(() => ConfigBuilder.GetReloadToken(), () =>
{
     Console.WriteLine("Reload");
});

### Expected behavior

both "./api.json" and "api.json" can fire onChange event

### Actual behavior

it seems like in PhysicalFilesWatcher.ReportChangeForMatchedEntries the _filePathTokenLookup save "./api.json" as key but try to match with "api.json"

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.net6.0
Microsoft.Extensions.Hosting 7.0.0-preview.5.22301.12
Microsoft.Extensions.Primitives 7.0.0-preview.5.22301.12

### Other information

_No response_

### PR DESCRIPTION
**Pull request by turnyanszki**
#if NET5_0_OR_GREATER is needed beacuse the project is targeting .net framwork for some reason (analyzers maybe).
All project should be refactored at this lib some point in my opinion.
fixes #71386

### CODE REVIEW

**Comment by radical** [comment]
@turnyanszki In the PR description, could you please describe the issue that you are fixing here? Also, add a short title for the PR. And reference any link for the issue, if there is one.

**Comment by danmoseley** [comment]
@Jozkee  could you please give an idea when you will be able to review? This change seems small.

**Comment by eerhardt** [comment]
@turnyanszki - looks like the new test you added to verify the bug fix is failing on .NET Framework. This is because you are only adding code for `#if NET5_0_OR_GREATER`. You need to also fix the bug for .NET Framework.

File: src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs [hunk 1]
```diff
@@ -175,6 +175,10 @@ private IChangeToken GetOrAddChangeToken(string pattern)
 
         internal IChangeToken GetOrAddFilePathChangeToken(string filePath)
         {
+#if NET5_0_OR_GREATER
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:178]:
```suggestion
#if NETCOREAPP
```

We only build this library on .NET Core for net6.0+. So the if can be simplified (we don't use lower version checks than what we build for).

**Comment by turnyanszki** [comment]
> @turnyanszki - looks like the new test you added to verify the bug fix is failing on .NET Framework. This is because you are only adding code for `#if NET5_0_OR_GREATER`. You need to also fix the bug for .NET Framework.

I dont want to implement the whole Path.GetFullPath method myself, and it does not exists in .NET Framework.
Any suggestions?

**Comment by eerhardt** [comment]
> I dont want to implement the whole Path.GetFullPath method myself, and it does not exists in .NET Framework.
Any suggestions?

Looking at `Path.GetFullPath`, the trickiest part is the `if (IsPathFullyQualified(path))` method:

https://github.com/dotnet/runtime/blob/878354a9d2c159cec527edc94bffdd3324570feb/src/libraries/System.Private.CoreLib/src/System/IO/Path.Unix.cs#L36-L51

This method is available on `netstandard2.0` or .NET Framework. However, looking at the callers of this method, it call comes from `CreateFileChildToken` here:

https://github.com/dotnet/runtime/blob/878354a9d2c159cec527edc94bffdd3324570feb/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs#L134-L146

Which verifies that the `filter` string isn't a rooted path. So I believe you can just Combine the two paths, and call GetFullPath on the result. That _should_ be the same on .NET Framework.

File: src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs [hunk 2]
```diff
@@ -175,6 +175,10 @@ private IChangeToken GetOrAddChangeToken(string pattern)
 
         internal IChangeToken GetOrAddFilePathChangeToken(string filePath)
         {
+#if NETCOREAPP
+            var absolutePath = Path.GetFullPath(filePath, _root);
+            filePath = absolutePath.Substring(_root.Length);
+#endif
```
Comment (Jozkee) [inline src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:181]: @eerhardt suggestion, this should work for .NET Framework too:

```suggestion
#if NETCOREAPP
            string absolutePath = Path.GetFullPath(filePath, _root);
#else
            string absolutePath = Path.GetFullPath(Path.Combine(_root, filePath));
#endif
            filePath = absolutePath.Substring(_root.Length);
```

File: src/libraries/Microsoft.Extensions.Configuration/tests/FunctionalTests/ConfigurationTests.cs [hunk 3]
```diff
@@ -987,6 +987,41 @@ void ReloadLoop()
             }
         }
 
+        [Fact]
+        public async Task ReloadingFileFiresOnChangeEvent()
```
Comment (Jozkee) [inline src/libraries/Microsoft.Extensions.Configuration/tests/FunctionalTests/ConfigurationTests.cs:991]: This test should be a `Theory` and you should put `./reload.json` and `reload.json` as `InlineData`.

**Review by jozkee — CHANGES_REQUESTED** [review]

**Pull request closed without merging**